ipa/ipa_libcloud.py

```diff
@@ -20,13 +20,51 @@
 # You should have received a copy of the GNU General Public License
 # along with this program.  If not, see <http://www.gnu.org/licenses/>.
 
+from libcloud.common.exceptions import BaseHTTPError
+
 from ipa.ipa_exceptions import LibcloudProviderException
 from ipa.ipa_provider import IpaProvider
 
 
 class LibcloudProvider(IpaProvider):
     """Provider class for testing images with libcloud."""
 
+    def _get_image(self):
+        """Retrieve NodeImage given the image id."""
+        try:
+            images = self.compute_driver.list_images()
```

Comment:
This is a big performance hit, a list of images in EC2 returns thousands of images. All that information needs to be sent over the wire consuming bandwidth, never mind that the code then has to fish through all those entries to do the comparison. I am not in favor of anything that include list_images(). IPA should trust the user input, use whatever image identifier the user provided and then try to launch an instance. If that fails we error out based on handling the "launch_instance" exception rather than fishing through a very large list of available images.

Reply:
It didn't see a huge hit in GCE or EC2. However, the drivers should each extend get_image. Unfortunately GCE is still using ex_get_image. Split it back up.